Added meshlink_free to API.

a-krebs opened the pull request:
Using meshlink_free on memory allocated by meshlink (rather than calling
free directly) resolves memory management problems on windows; the dll
that allocated the memory must also free it, otherwise heap check
assertions fail.

a-krebs committed:
Added meshlink_free to API.

Using meshlink_free on memory allocated by meshlink (rather than calling
free directly) resolves memory management problems on windows; the dll
that allocated the memory must also free it, otherwise heap check
assertions fail.

everHannes commented:
I'm not sure on linux, but on windows deallocation across dll boundaries really should be avoided, given each dll might have it's own runtime, see http://blogs.msdn.com/b/oldnewthing/archive/2006/09/15/755966.aspx

I think the meshlink_free wrapper is ok, although it requires some additional care.
To simplify this the return of the c++ interface could already be wrapped to call meshlink_free on release.

Another common way I've seen in c code is to require memory allocation prior to the function call and return what size is used or required if too small. That way same buffer could be reused for multiple calls and the caller can also std::move the memory for later usage.
For c coding I think that is what I would prefer.

a-krebs committed:
Change C++ API to encapsulate meshlink_free.

Calls to meshlink_free where only required for char* return values.
These can be replaced by std::string, and then calls to meshlink_free
are not needed at all when using the C++ API.

a-krebs committed:
Update examples to use new C++ API.
